pkg/idp/provider/simpleprovider/simpleprovider.go

```diff
@@ -0,0 +1,187 @@
+	}, nil
+}
+
+func (s *SimpleProvider) GetName() string {
+	return s.Name
+}
+
+func (s *SimpleProvider) GetProviderName() string {
+	return ProviderName
+}
+
+func (s *SimpleProvider) GetType() string {
+	return s.Type
+}
+
+func (s *SimpleProvider) GetOwner() string {
+	return s.Owner
+}
+
+func (s *SimpleProvider) CreateOrUpdateApp(config provider.AppConfig, ctx context.Context, oldConnector dex.Connector) (provider.ProviderApp, error) {
+	// Inject the redirect URI into the connector config
+	connectorConfig := s.injectRedirectURI(config.RedirectURI)
+
+	// validate that the connector type is an existing type and that connector config is valid for the given connector type
+	err := s.validateConnectorConfig(connectorConfig)
+	if err != nil {
+		return provider.ProviderApp{}, microerror.Mask(err)
+	}
+
+	return provider.ProviderApp{
+		Connector: dex.Connector{
+			Type:   s.ConnectorType,
+			ID:     s.Name,
+			Name:   s.Description,
+			Config: connectorConfig,
+		},
+		SecretEndDateTime: time.Now().AddDate(0, 6, 0),
+	}, nil
+}
+
+func (s *SimpleProvider) DeleteApp(name string, ctx context.Context) error {
+	return nil
+}
+
+func (s *SimpleProvider) GetCredentialsForAuthenticatedApp(config provider.AppConfig) (map[string]string, error) {
+	s.Log.Info(fmt.Sprintf("No new credentials will be created for the %s provider because it does not allow dex-operator access.", ProviderName))
+	return map[string]string{}, nil
+}
+func (s *SimpleProvider) CleanCredentialsForAuthenticatedApp(config provider.AppConfig) error {
+	return nil
+}
+
+func (s *SimpleProvider) DeleteAuthenticatedApp(config provider.AppConfig) error {
+	return nil
+}
+
+func (s *SimpleProvider) injectRedirectURI(redirectURI string) string {
+	config := s.ConnectorConfig
+
+	if !usesRedirectURI(s.ConnectorType) {
+		return config
+	}
+
+	// if a redirect URI is already defined in the config, we replace it with the new one
+	if strings.Contains(config, "redirectURI") {
+		regexp := regexp.MustCompile(`redirectURI:.*`)
+		config = regexp.ReplaceAllString(config, fmt.Sprintf("redirectURI: %s", redirectURI))
+	} else {
+		config = fmt.Sprintf("%s\nredirectURI: %s", config, redirectURI)
+	}
+	return config
+}
+
+func (s *SimpleProvider) validateConnectorConfig(connectorConfig string) error {
+	// validate that the connector type is an existing type
+	f, ok := server.ConnectorsConfig[s.ConnectorType]
+	if !ok {
+		return microerror.Maskf(invalidConfigError, "Unknown connector type %q", s.ConnectorType)
+	}
+
+	// validate that connector config is valid for the given connector type
+	connConfig := f()
+	if len(connectorConfig) != 0 {
+		data := []byte(connectorConfig)
+		if err := yaml.Unmarshal(data, connConfig); err != nil {
+			return microerror.Maskf(invalidConfigError, "Parse connector config: %v", err)
+		}
+	}
+
+	return nil
+}
+
+func usesRedirectURI(connectorType string) bool {
+	switch connectorType {
+	case "ldap", "authproxy", "atlassian-crowd", "keystone":
```

Comment:
These types do not use callback so it would cause an error to inject redirectURI.